File: code/ex368.java
import java.util.*;

class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        if (nums.length == 0)    return new ArrayList<Integer>();
        
        Arrays.sort(nums);
        
        // Use dp to find the max length of the subset and the index of max number in the subset
        int max = 0, maxIndex = 0;
        int[] dp = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            for (int j = i - 1; j >= 0; j--)
                if (nums[i] % nums[j] == 0)
                    dp[i] = Math.max(dp[i], dp[j] + 1);

            if (dp[i] == 0)    dp[i] = 1;
            
            if (dp[i] > max) {
                max = dp[i];
                maxIndex = i;
            }
        }
        
        List<Integer> res = new ArrayList<>();
        // Find all the numbers in the array which can be divisible by nums[maxIndex] and add them to the list
        int dividend = nums[maxIndex], restLen = max;
        for (int i = maxIndex; i >= 0 && restLen >= 0; i--)
            if (dp[i] == restLen && dividend % nums[i] == 0) {
                res.add(nums[i]);
                dividend = nums[i];
                restLen--;
            }
                
        
        return res;
    }
}

File: code/ex368.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n = nums.size();    if (n == 0)    return {};
        sort(nums.begin(), nums.end());
        
        int MAX = 1, end = 0;
        vector<int> dp(n, 1), prev(n);
        for (int i = 0; i < n; i++) {
            prev[i] = i;
            for (int j = 0; j < i; j++) {
                if (nums[i] % nums[j] == 0) {
                    if (dp[j] + 1 > dp[i]) {
                        dp[i] = dp[j] + 1;
                        prev[i] = j;
                    }
                }
            }
            if (dp[i] > MAX) {
                MAX = dp[i];
                end = i;
            }
        }
        
        // Get subset
        vector<int> res;
        while (1) {
            res.push_back(nums[end]);
            if (prev[end] == end)    break;
            end = prev[end];
        }
        
        return res;
    }
};